File: Java/dependency-injection/_02_control-freak/src/main/java/com/github/grzesiek_galezowski/DependencyInjection/_1_ControlFreak/Inbound/BinaryUdpInbound.java
package com.github.grzesiek_galezowski.DependencyInjection._1_ControlFreak.Inbound;

import com.github.grzesiek_galezowski.DependencyInjection._1_ControlFreak.Core.AcmeProcessingWorkflow;
import com.github.grzesiek_galezowski.DependencyInjection._1_ControlFreak.Interfaces.AcmeMessage;

public class BinaryUdpInbound {
    private AcmeProcessingWorkflow workflow;
    private final UdpSocket socket;
    private final BinaryParsing parsing;

    public BinaryUdpInbound() {
      socket = new UdpSocket();
      parsing = new BinaryParsing();
    }

    public void setDomainLogic(AcmeProcessingWorkflow processingWorkflow) {
      workflow = processingWorkflow;
    }

    public void startListening() {
      byte[] frameData = new byte[100];
      while (socket.receive(frameData)) {
        AcmeMessage message = parsing.resultFor(frameData);
        if (message != null) {
          if (workflow != null) {
            workflow.applyTo(message);
          }
        }
      }
    }
  }

File: CSharp/DependencyInjection/DependencyInjection/_1_ControlFreak/Inbound/BinaryUdpInbound.cs
using DependencyInjectionBefore._1_ControlFreak.Core;

namespace DependencyInjectionBefore._1_ControlFreak.Inbound
{
  class BinaryUdpInbound
  {
    private AcmeProcessingWorkflow _processingWorkflow;
    private readonly UdpSocket _socket;
    private readonly BinaryParsing _parsing;

    public BinaryUdpInbound()
    {
      _socket = new UdpSocket();
      _parsing = new BinaryParsing();
    }

    public void SetDomainLogic(AcmeProcessingWorkflow processingWorkflow)
    {
      _processingWorkflow = processingWorkflow;
    }
    
    public void StartListening()
    {
      byte[] frameData;
      while (_socket.Receive(out frameData))
      {
        var message = _parsing.ResultFor(frameData);
        if (message != null)
        {
          if (_processingWorkflow != null)
          {
            _processingWorkflow.ApplyTo(message);
          }
        }
      }
    }
  }
}